providers/curation/github.js

```diff
@@ -144,15 +145,60 @@ class GitHubCurationService {
     return serviceGithub.repos.createFile(fileBody)
   }
 
+  // Return an array of valid patches that exist
+  // and a list of definitions that do not exist in the store
+  async _validateComponentDefinitionExists(patches) {
+    const coordinatesWithRevisions = flatten(
+      map(patches, patch =>
+        map(patch.revisions, (_, key) => {
+          return Object.assign({}, patch.coordinates, {
+            revision: key
+          })
+        })
+      )
+    )
+
+    const validDefinitions = await this.definitionService.listAll(coordinatesWithRevisions)
+    const validPatches = Object.assign([], patches)
+    const notFoundDefinitions = []
+    map(patches, (patch, index) =>
+      map(
+        patch.revisions,
+        throat(10, async (_, key) => {
+          const coordinates = Object.assign({}, patch.coordinates, {
+            revision: key
+          })
+          if (validDefinitions.findIndex(definition => isEqual(definition, coordinates)) === -1) {
+            delete validPatches[index].revisions[key]
+            if (isEmpty(validPatches[index].revisions)) delete validPatches[index]
+            notFoundDefinitions.push(
+              `Definition ${EntityCoordinates.fromObject(coordinates).toString()} has not been found`
+            )
+          }
+        })
+      )
+    )
+    return { validPatches: compact(validPatches), notFoundDefinitions }
+  }
+
   async addOrUpdate(userGithub, serviceGithub, info, patch) {
     const { owner, repo, branch } = this.options
     const masterBranch = await serviceGithub.repos.getBranch({ owner, repo, branch: `refs/heads/${branch}` })
     const sha = masterBranch.data.commit.sha
     const prBranch = await this._getBranchName(info)
     await serviceGithub.gitdata.createReference({ owner, repo, ref: `refs/heads/${prBranch}`, sha })
+    const { validPatches, notFoundDefinitions } = await this._validateComponentDefinitionExists(patch.patches)
+    const validResult = { errors: notFoundDefinitions }
+    if (validPatches.length === 0) {
```

Comment:
Good question here is whether we should succeed at all if there are failures. My first cut at this is to say fail it all if any one of them are bad. That is part of the value of putting the check higher up in the code. We can do that check for all before writing anything

Here I would throw. In the route we can catch and return a 400

Reply:
I've understand your point, but from a contributor point of view i think it couldn't be worth.
Thinking about a contribution made of hundreds of components, if only one them is not existing then the whole PR should fail?
What we've done here is to allow all the valid definitions to be contributed, and return a message for those that are not existing.
We could definitely go back and stop the entire process if something is wrong, we just want to be sure that is the expected behavior during a contribution process.

Reply:
as discussed in the call today, we decided to fail the entire call if there are any problems. The API caller should get a 400 and hopefully some indication as to what was wrong.